### PR TITLE
Setting System Feature

### PR DESCRIPTION
**Pull request by afshin**
This PR introduces a setting system for JupyterLab. There are a few core concepts:
* Settings are stored as JSON blobs tied/namespaced to plugins.
* Each plugin's settings have two tiers: `user` and `system`. The `system` JSON represents the default values and is **read-only**. The `user` values are editable by the end user by opening the settings editor from the `Help` menu.

### Change list
- [x] Add the `IDatastore` interface to describe a low-level read/write database. `IStateDB` and `ISettingRegistry` both depend on it. This simplifies swapping out back-ends for state and setting databases, and others if they arise.
- [x] Make `InstanceTracker` independent of `@jupyterlab/application`.
- [x] Create a setting registry extension: `ISettingRegistry`, defined in `@jupyterlab/coreutils`.
- [x] Create a setting system editor.

### CODE REVIEW

File: packages/fileeditor-extension/src/index.ts [hunk 1]
```diff
@@ -107,25 +106,18 @@ function activate(app: JupyterLab, registry: IDocumentRegistry, restorer: ILayou
...
+    if (!wordWrap) {
+      commands.execute(CommandIDs.wordWrap);
     }
-    if (typeof settings['wordWrap'] === 'string') {
-      commands.execute(CommandIDs.wordWrap, settings);
-    }
-    if (typeof settings['lineNumbers'] === 'string') {
-      commands.execute(CommandIDs.lineNumbers, settings);
+
+    lineNumbers = !!settings.get('lineNumbers');
```
Comment (blink1073) [inline packages/fileeditor-extension/src/index.ts]: noice

**Comment by ellisonbg** [comment]
Let us know when this is ready for review ;-)

**Comment by afshin** [comment]
@ellisonbg I will let you know. And I'll write it up so that it's not just a big blob of code.

File: packages/apputils-extension/src/index.ts [hunk 2]
```diff
@@ -104,6 +104,21 @@ const palettePlugin: JupyterLabPlugin<ICommandPalette> = {
 
 
 /**
+ * The default layout restorer provider.
+ */
+const settingPlugin: JupyterLabPlugin<ISettingRegistry> = {
+  id: 'jupyter.services.setting-registry',
+  requires: [IStateDB],
+  activate: (app: JupyterLab, state: IStateDB) => {
+    const registry = new SettingRegistry();
```
Comment (blink1073) [inline packages/apputils-extension/src/index.ts]: Use the statedb here?

Comment (afshin) [inline packages/apputils-extension/src/index.ts]: This happens automatically inside the `SettingRegistry` constructor:

```typescript
/**
 * Create a new setting registry.
 */
constructor(options: SettingRegistry.IOptions = { datastore: null }) {
  const namespace = 'jupyter.db.settings';
  this._datastore = options.datastore || new StateDB({ namespace });
}
```

Comment (blink1073) [inline packages/apputils-extension/src/index.ts]: What I mean is that the state instance should be used (`new SettingRegistry({ datastore: state })`)

Comment (blink1073) [inline packages/apputils-extension/src/index.ts]: Either that or the plugin shouldn't depend on a statedb.

Comment (afshin) [inline packages/apputils-extension/src/index.ts]: Ah right, good point. I originally did it that way and then I forgot to remove `IStateDB` as a requirement when I switched it.

**Comment by afshin** [comment]
FYI @ellisonbg, I had some merge conflicts when I rebased, so while I was resolving them, I also refactored this file which may affect work you want to do: https://github.com/jupyterlab/jupyterlab/pull/2157/commits/e7dbedce921d5a329ce1f9ab3fa8eaab55add8c5

**Comment by ellisonbg** [comment]
@vineetgorhe would also affect you if you dive into the markdown stuff again.

File: packages/apputils-extension/src/index.ts [hunk 3]
```diff
@@ -101,6 +104,17 @@ const palettePlugin: JupyterLabPlugin<ICommandPalette> = {
 
 
 /**
+ * The default layout restorer provider.
```
Comment (blink1073) [inline packages/apputils-extension/src/index.ts]: incorrect description

**Comment by mpacer** [comment]
Just a note about an example of a SublimeText Package that has an extremely complicated settings format with tonnes of documentation: https://github.com/SublimeText/LaTeXTools.

It's not the case that we should design for, but it's something that is likely to come up if people are doing complicated publishing oriented tasks like writing papers/reports.

**Comment by blink1073** [comment]
Looks and works great!

**Pull request merged**